build-system/runner/src/org/ampproject/AmpPass.java

```diff
@@ -100,26 +99,36 @@ public AmpPass(AbstractCompiler compiler, boolean isProd,
    * and a child STRING "assert" (or any other signature from Set<String> value)
    */
   private boolean isCallRemovable(Node n) {
-    if (n != null && n.isCall()) {
-      Node grandparent = n.getGrandparent();
-      // We want to make sure an `Log.prototype` method is actually invoked and not just
-      // a pointer to it.
-      if (grandparent == null || !grandparent.isCall()) {
-        return false;
-      }
-      Node callname = n.getFirstChild();
-      Node getprop = n.getNext();
-      if (callname != null && callname.isGetProp()) {
-       Set<String> methodCallNames = stripTypeSuffixes
-          .get(callname.getQualifiedName());
-       if (methodCallNames != null) {
-         for (String methodCallName : methodCallNames) {
-           if (getprop != null && getprop.isString() &&
-               methodCallName == getprop.getString()) {
-             return true;
-           }
-         }
-       }
+    if (n == null || !n.isCall()) {
```

Comment:
Not sure whether it applies here, but one way to generally solve the type of problems is to make your compiler pass 2 phases:

1. Collect nodes to remove (add to list).
2. Iterate over nodes to remove and remove them.

This avoids having to think about how in-place changes affect the traversal.